gh-91243: Document Required and NotRequired (#93173)
python/cpython: Doc/library/typing.rst

   .. versionadded:: 3.8

.. data:: Required

.. data:: NotRequired

>    A special typing construct that marks individual keys of a :class:`TypedDict`

Reviewer: There are two

```suggestion
   Special typing constructs that mark individual keys of a :class:`TypedDict`
```

   as either required or non-required respectively.

>    For more information, see

Reviewer: You've added some good docs in this file; we should link to those more prominently than the PEP, which may grow out of date since it is now a historical document.

```suggestion
   For more information, see :class:`TypedDict` and
```

>    :pep:`655` (Marking individual TypedDict items as required or potentially-missing).

Reviewer: ```suggestion
   :pep:`655` ("Marking individual TypedDict items as required or potentially-missing").
```

      Point2D = TypedDict('Point2D', {'in': int, 'x-y': int})

   By default, all keys must be present in a ``TypedDict``. It is possible to

>    mark individual keys as non-required using ``NotRequired``::

Reviewer: ```suggestion
   mark individual keys as non-required using :data:`NotRequired`::
```

   ``True`` as the value of the ``total`` argument. ``True`` is the default,
   and makes all items defined in the class body required.

   Individual keys of a ``total=False`` ``TypedDict`` can be marked as

>    required using ``Required``::

Reviewer: ```suggestion
   required using :data:`Required`::
```

      ``Point2D.__required_keys__`` and ``Point2D.__optional_keys__`` return
      :class:`frozenset` objects containing required and non-required keys, respectively.

      Keys marked with :data:`Required` will always appear in ``__required_keys__``

>       and keys marked with ``NotRequired`` will always appear in ``__optional_keys__``.

Reviewer: ```suggestion
      and keys marked with :data:`NotRequired` will always appear in ``__optional_keys__``.
```

      For backwards compatibility with Python 3.10 and below,
      it is also possible to use inheritance to declare both required and
      non-required keys in the same ``TypedDict`` , declaring a ``TypedDict`` with one value
      for the ``total`` argument and then inheriting it from another ``TypedDict`` with
      a different value for ``total``.

>       Usage::

Reviewer: ```suggestion
      non-required keys in the same ``TypedDict`` . This is done by declaring a
      ``TypedDict`` with one value for the ``total`` argument, and then
      inheriting from it in another ``TypedDict`` with a different value for
      ``total``::
```

      ``Point2D.__required_keys__`` and ``Point2D.__optional_keys__`` return
      :class:`frozenset` objects containing required and non-required keys, respectively.

      Keys marked with ``Required`` will always appear in ``__required_keys__``
      and keys marked with ``NotRequired`` will always appear in ``__optional_keys__``.

>       It is also possible to use mixed inheritance to declare both required and

Reviewer 2: Maybe we should just drop this example, there's no reason to use it any more.

Reviewer: It's still useful for backwards compatibility, if users don't want to have `typing_extensions` as a dependency.

```suggestion
      For backwards compatibility, it is also possible to use mixed inheritance to declare both required and
```

Author: I did add the more-specific prelude:

> For backwards compatibility with Python 3.10 and below, [...]

      ``Point2D.__required_keys__`` and ``Point2D.__optional_keys__`` return
      :class:`frozenset` objects containing required and non-required keys, respectively.

>       Keys marked with ``Required`` will always appear in ``__required_keys__``

Reviewer: ```suggestion
      Keys marked with :data:`Required` will always appear in ``__required_keys__``
```

      and keys marked with ``NotRequired`` will always appear in ``__optional_keys__``.

      For backwards compatibility with Python 3.10 and below,

>       it is also possible to use mixed inheritance to declare both required and

Reviewer: I'm not sure what "mixed inheritance" means, and the example below makes the meaning of your paragraph clear anyway

```suggestion
      it is also possible to use inheritance to declare both required and
```

   .. versionadded:: 3.8

   .. versionchanged:: 3.11

>       Added support for marking individual keys as ``Required`` or ``NotRequired``.

Reviewer: ```suggestion
      Added support for marking individual keys as :data`Required` or :data:`NotRequired`.
```
